futures-util/src/future/shared.rs

```diff
@@ -156,153 +203,76 @@ impl<F> Future for Shared<F>
-                let waker = Waker::from(self.inner.notifier.clone());
-                let mut cx = task::Context::new(data, &waker, cx.executor());
-                f.poll(&mut cx)
+                if let FutureOrOutput::Future(future) = &mut *this.inner.future_or_output.get() {
+                    PinMut::new_unchecked(future).poll(&mut cx)
+                } else {
+                    unreachable!()
+                }
             };
             match res {
-                Ok(Async::Pending) => {
+                Poll::Pending => {
                     // Not ready, try to release the handle
-                    match self.inner.notifier.state.compare_and_swap(POLLING, IDLE, SeqCst) {
+                    match this.inner.notifier.state.compare_and_swap(POLLING, IDLE, SeqCst) {
                         POLLING => {
                             // Success
-                            return Ok(Async::Pending);
+                            return Poll::Pending;
                         }
                         REPOLL => {
                             // Gotta poll again!
-                            let prev = self.inner.notifier.state.swap(POLLING, SeqCst);
+                            let prev = this.inner.notifier.state.swap(POLLING, SeqCst);
                             assert_eq!(prev, REPOLL);
                         }
                         _ => unreachable!(),
                     }
-
                 }
-                Ok(Async::Ready(i)) => {
+                Poll::Ready(i) => {
                     unsafe {
-                        (*self.inner.result.get()) = Some(Ok(SharedItem { item: Arc::new(i) }));
+                        *this.inner.future_or_output.get() = FutureOrOutput::Output(Arc::new(i));
                     }
-
-                    break;
-                }
-                Err(e) => {
-                    unsafe {
-                        (*self.inner.result.get()) = Some(Err(SharedError { error: Arc::new(e) }));
-                    }
-
                     break;
                 }
             }
         }
 
-        self.complete();
-        unsafe { self.clone_result().map(Async::Ready) }
+        unsafe {
+            this.complete();
+            Poll::Ready(this.clone_result())
+        }
     }
 }
 
 impl<F> Clone for Shared<F> where F: Future {
     fn clone(&self) -> Self {
-        let next_clone_id = self.inner.next_clone_id.fetch_add(1, SeqCst);
-
         Shared {
             inner: self.inner.clone(),
-            waiter: next_clone_id,
+            waker_key: NULL_WAKER_KEY,
         }
     }
 }
 
 impl<F> Drop for Shared<F> where F: Future {
     fn drop(&mut self) {
-        let mut waiters = self.inner.notifier.waiters.lock().unwrap();
-        waiters.remove(&self.waiter);
+        if self.waker_key != NULL_WAKER_KEY {
+            if let Ok(mut wakers) = self.inner.notifier.wakers.lock() {
+                wakers.remove(self.waker_key);
```

Comment:
BTW `Slab::remove()` doesn't actually shrink the `Vec` inside the `Slab`. See suggestion above about wrapping the `Slab` in an `Option`,

Reply:
@MajorBreakfast Yeah, I opened an issue about fixing `Slab::remove` long ago-- see https://github.com/carllerche/slab/issues/38